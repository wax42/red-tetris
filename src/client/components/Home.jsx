import React from "react";
import { withRouter } from "react-router-dom";

let roomName = "";
let playerName = "";
let url = "";

//TODO validation
function getRoomName(event) {
  console.log("test player name");
  roomName = event.target.value;
  console.log(roomName);
}

function getPlayerName(event) {
  console.log("test player name");
  playerName = event.target.value;
  console.log(playerName);
}
// return <Redirect to='/dashboard' />

const onSubmit = history => {
  /* console.log("On submit");
  if (playerName.length > 3) return <Link to="/app">Valider</Link>;
  else return <div>error</div>; */
  console.log(roomName);
  console.log(playerName);
  if (roomName.length < 3 || playerName.length < 3) console.error("name < 3 ");
  else 
  history.push(`/#${roomName}[${playerName}]`);
};

const Button = withRouter(({ history }) => (
  <button
    type="button"
    onClick={() => {
      onSubmit(history);
    }}
  >
    Click Me!
  </button>
));

const Home = () => {
  return (
    <div className="home">
      <h1>Red Tetris</h1>
      <div className="home-field">
        Room name
        <input type="text" name="room-name" onChange={e => getRoomName(e)} />
      </div>
      <div className="home-field">
        Player name
        <input
          type="text"
          name="player-name"
          onChange={e => getPlayerName(e)}
        />
      </div>
      <div>
        <Button />
      </div>
    </div>
  );
};

export default Home;
