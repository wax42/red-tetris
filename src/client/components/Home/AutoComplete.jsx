import React from "react";
import deburr from "lodash/deburr";
import Autosuggest from "react-autosuggest";
import match from "autosuggest-highlight/match";
import parse from "autosuggest-highlight/parse";
import TextField from "@material-ui/core/TextField";
import Paper from "@material-ui/core/Paper";
import MenuItem from "@material-ui/core/MenuItem";
import { makeStyles } from "@material-ui/core/styles";
import _ from "lodash";

let suggestions = [];

export const renderInputComponent = inputProps => {
  const { classes, inputRef = () => {}, ref, ...other } = inputProps;

  return (
    <TextField
      margin="normal"
      variant="outlined"
      InputProps={{
        inputRef: node => {
          ref(node);
          inputRef(node);
        },
        classes: {
          input: classes.input
        }
      }}
      {...other}
    />
  );
};

export const renderSuggestion = (suggestion, { query, isHighlighted }) => {
  const matches = match(suggestion.label, query);
  const parts = parse(suggestion.label, matches);

  return (
    <MenuItem selected={isHighlighted} component="div">
      <div>
        {parts.map(part => (
          <span key={part.text} style={{ fontWeight: part.highlight ? 500 : 400 }}>
            {part.text}
          </span>
        ))}
      </div>
    </MenuItem>
  );
};

export const getSuggestions = value => {
  const inputValue = deburr(value.trim()).toLowerCase();
  const inputLength = inputValue.length;
  let count = 0;

  return inputLength === 0
    ? []
    : suggestions.filter(suggestion => {
        const keep = count < 5 && suggestion.label.slice(0, inputLength).toLowerCase() === inputValue;

        if (keep) {
          count += 1;
        }

        return keep;
      });
};

export function getSuggestionValue(suggestion) {
  return suggestion.label;
}

const useStyles = makeStyles(theme => ({
  root: {
    height: 250,
    flexGrow: 1
  },
  container: {
    position: "relative"
  },
  suggestionsContainerOpen: {
    position: "absolute",
    zIndex: 1,
    marginTop: theme.spacing(1),
    left: 0,
    right: 0
  },
  suggestion: {
    display: "block"
  },
  suggestionsList: {
    margin: 0,
    padding: 0,
    listStyleType: "none"
  },
  divider: {
    height: theme.spacing(2)
  }
}));

export const AutoComplete = ({ roomName, setRoomName, listRooms }) => {
  const classes = useStyles();

  suggestions = _.map(listRooms, value => {
    return { label: value };
  });

  const [stateSuggestions, setSuggestions] = React.useState([]);

  const handleSuggestionsFetchRequested = ({ value }) => {
    setSuggestions(getSuggestions(value));
  };

  const handleSuggestionsClearRequested = () => {
    setSuggestions([]);
  };

  const handleChange = () => (event, { newValue }) => {
    setRoomName(newValue);
  };

  const autosuggestProps = {
    renderInputComponent,
    suggestions: stateSuggestions,
    onSuggestionsFetchRequested: handleSuggestionsFetchRequested,
    onSuggestionsClearRequested: handleSuggestionsClearRequested,
    getSuggestionValue,
    renderSuggestion
  };

  return (
    <Autosuggest
      {...autosuggestProps}
      inputProps={{
        classes,
        id: "react-autosuggest-simple",
        label: "Room",
        placeholder: "Enter a room name",
        value: roomName,
        onChange: handleChange()
      }}
      theme={{
        container: classes.container,
        suggestionsContainerOpen: classes.suggestionsContainerOpen,
        suggestionsList: classes.suggestionsList,
        suggestion: classes.suggestion
      }}
      renderSuggestionsContainer={options => (
        <Paper {...options.containerProps} square>
          {options.children}
        </Paper>
      )}
    />
  );
};
