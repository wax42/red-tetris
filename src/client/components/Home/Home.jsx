import React, { useEffect, useState, createRef } from "react";
import { FaHome, FaIdBadge } from "react-icons/fa";
import { connect } from "react-redux";
import { actionCreateRoom } from "../../actions/actionsRedux";
import { Button, TextField, Snackbar } from "@material-ui/core";
import { AutoComplete } from "./AutoComplete";
import ERROR from "../../../common/error";

export const buttonCreateRoom = (action, setStateError, roomName, playerName) => {
  if (roomName.length < 3 && roomName.length > 12) {
    setStateError(ERROR.ROOMNAME_INVALID_LENGTH);
  } else if (playerName.length < 3 && playerName.length > 12) {
    setStateError(ERROR.PLAYERNAME_INVALID_LENGTH);
  } else if (/^[A-z0-9]+$/.test(roomName) === false) {
    setStateError(ERROR.ROOMNAME_INVALID);
  } else if (/^[A-z0-9]+$/.test(playerName) === false) {
    setStateError(ERROR.PLAYERNAME_INVALID);
  } else {
    window.location.hash = `#${roomName}[${playerName}]`;
    action(roomName, playerName);
  }
};

const mapStateToProps = _state => {
  const listPlayers = _state.listPlayers;
  const listRooms = _state.listRooms;
  return { listPlayers, listRooms };
};

export const HomeCpt = ({ listPlayers, listRooms, theme, error, actionCreateRoom }) => {
  const [stateError, setStateError] = useState(error);
  const [playerName, setPlayerName] = useState("");
  const [roomName, setRoomName] = useState("");

  console.log(stateError);
  return (
    <div>
      <div className="home-title">Red Tetris</div>
      <div className="home">
        <form className="form-home">
          <div className="home-field">
            <div className="home-field-title">
              <FaIdBadge className="home-icon" />
              <div>Room name</div>
            </div>
            <AutoComplete listRooms={listRooms} roomName={roomName} setRoomName={setRoomName} />
          </div>
          <div className="home-field">
            <div className="home-field-title">
              <FaHome className="home-icon" />
              <div>Player name</div>
            </div>
            <TextField
              color="primary"
              label="Player"
              className={"homeTextField"}
              // value={playerName}
              onChange={e => setPlayerName(e.target.value)}
              margin="normal"
              variant="outlined"
            />
          </div>

          <Button
            color="secondary"
            size="large"
            variant="outlined"
            onClick={() => buttonCreateRoom(actionCreateRoom, setStateError, roomName, playerName)}
          >
            Start
          </Button>
        </form>
      </div>
      <Snackbar
        anchorOrigin={{ vertical: "bottom", horizontal: "center" }}
        key="key-msg"
        open={stateError !== null}
        onClose={() => setStateError(null)}
        ContentProps={{
          "aria-describedby": "message-id"
        }}
        message={<span id="message-id">{stateError}</span>}
      />

      <div className="error-container">{stateError}</div>

      <div className="bird-container">
        <div className="bird">
          <div className="bird-beak2" />
          <div className="bird-beak1" />
          <div className="bird-body" />
          <div className="bird-eye" />
          <div className="bird-pupil" />
          <div className="bird-wing" />
          <div className="bird-hair" />
          RED TETRIS by ppichier & vguerand
        </div>
      </div>
    </div>
  );
};

const Home = connect(
  mapStateToProps,
  { actionCreateRoom }
)(HomeCpt);

export default Home;
