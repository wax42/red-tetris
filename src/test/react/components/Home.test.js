import React from "react";
import Home, {
  HomeCpt,
  getRoomName,
  getPlayerName,
  buttonCreateRoom
} from "../../../client/components/Home/Home";
import { shallow } from "enzyme";

describe("HOME.JSX", () => {
  it("should render self and subcomponent Home", () => {
    const setup = () => {
      const props = {
        state: {},
        actionCreateRoom: jest.fn(),
        error: ""
      };

      const enzymeWrapper = shallow(<HomeCpt />);

      return {
        props,
        enzymeWrapper
      };
    };
    const { enzymeWrapper } = setup();
    expect(enzymeWrapper.find("Title")).toBeTruthy();
    expect(enzymeWrapper.find("Info")).toBeTruthy();
  });

  it("should set the value of the roomName", () => {
    let roomName = "";
    const event = {
      target: {
        value: "room"
      }
    };
    getRoomName(event);
  });

  it("should set the value of the playerName", () => {
    let playerName = "";
    const event = {
      target: {
        value: "player"
      }
    };
    getPlayerName(event);
  });

  it("should set stateError if roomName < 3", () => {
    let roomName = "ab";
    let stateError = "";
    const action = jest.fn();
    const setStateError = msg => {
      stateError = msg;
    };
    buttonCreateRoom(
      action,
      setStateError("Room name should have 3 characters at least")
    );
  });
});
