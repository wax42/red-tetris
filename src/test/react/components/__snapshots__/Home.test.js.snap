// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HOME.JSX should render self and subcomponent Home 1`] = `ShallowWrapper {}`;
